modules/ardrivo/src/Entrypoint.cxx

```diff
@@ -16,15 +16,70 @@
  *
  */
 
+#include <algorithm>
+#include <memory>
+#include <range/v3/algorithm/for_each.hpp>
+#include <range/v3/view/zip.hpp>
 #include "BoardData.hxx"
 #include "BoardInfo.hxx"
 #include "Entrypoint.hxx"
+#include "utility.hxx"
 
-BoardData* board_data;
-const BoardInfo* board_info;
+BoardData* board_data = nullptr;
+const BoardInfo* board_info = nullptr;
+
+void merge_boards(BoardData old_board, BoardData& new_board) noexcept;
 
 bool init(BoardData* brd, const BoardInfo* info) {
+    if (board_data) {
+        auto defer_del_bd = std::unique_ptr<BoardData>{board_data};
+        auto defer_del_bi = std::unique_ptr<const BoardInfo>{board_info};
+        merge_boards(std::move(*board_data), *brd);
+    }
     board_data = brd;
     board_info = info;
     return true;
+}
+
+void maybe_init() noexcept {
+    if (board_data)
+        return;
+
+    // over-allocate everything to handle possible IO at dynamic-init
+    auto loc_board_data = std::make_unique<BoardData>();
+    loc_board_data->silence_errors = true;
+    loc_board_data->start_time = std::chrono::steady_clock::now();
+    loc_board_data->pin_modes = std::vector<std::atomic_uint8_t>(255);
+    loc_board_data->digital_pin_values = std::vector<std::atomic_bool>(255);
+    loc_board_data->analog_pin_values = std::vector<std::atomic_uint16_t>(255);
+    loc_board_data->i2c_buses = std::vector<I2cBus>(64);
+    loc_board_data->uart_buses = std::vector<UartBus>(64);
+    loc_board_data->interrupts_handlers = decltype(board_data->interrupts_handlers)(255);
+    board_data = loc_board_data.release();
+
+    auto loc_board_info = std::make_unique<BoardInfo>();
+    loc_board_info->pins_caps.resize(255);
+    ranges::for_each(loc_board_info->pins_caps, [](PinCapability& pin){ pin.digital_in = pin.digital_out = true; }); // default to all digital-able
+    // Protocol-pins need not be set as we cannot do pin-locking detection during dynamic-init anyway
+
+    ::board_info = loc_board_info.release();
```

Comment:
Cant this just be just `board_info` without the `::`?

Reply:
It can, but I want to make it clear that it's global scope